Standardising truncation of rdcon and stride

StuartBenjamin opened the pull request:
A short pull request that takes the q-based truncation logic from stride (qhigh and qlow) and adds it to rdcon. 

Old incorrect explanation: (updated comments below)
[I also included the reform_eq_with_psilim option, which if set to true, will re-spline the equilibrium using the limits set by qhigh, qlow, and sasflag. To keep backwards compatibility in rdcon, reform_eq_with_psilim is set to false by default, and the sasflag truncation does not override reform_eq_with_psilim=false. However, if qlow sets a new inner truncation point by redefining psilow, then reform_eq_with_psilim is forced true. Let me know if you think this is the right way to do it - the motivation came from not being sure how gal.f treats cases where psilow and psihigh don't correspond with the equilibrium spline bounds.]

StuartBenjamin committed:
initialising truncation control terms qhigh, qlow and reform_eq_with_psilim

StuartBenjamin committed:
adding new truncation control terms to rdcon_control namelist

StuartBenjamin committed:
adding new section in dcon.f for equilibrium reformation. Calling sing_lim and sing_min functions (sing_min to be added)

StuartBenjamin committed:
adding sing_min function (code is identical to stride version)

StuartBenjamin committed:
sing_min description updates

StuartBenjamin committed:
qlim now set by min(qmax,qhigh), same as in stride

StuartBenjamin committed:
sas_flag now has if, then logic (same as in stride)

StuartBenjamin committed:
psilim newton iterator now has if then logic, same as in stride

StuartBenjamin committed:
reforming equilibrium based new qhigh, qlow and sasflag limits

StuartBenjamin committed:
adding warning that using qlow will the reformation of the equilibrium spline

StuartBenjamin committed:
Adding description of new truncation terms to rdcon.in. Warnings regarding the dependence of qhigh and qlow on reform_eq_with_psilim are an attempt to try and keep backwards compatibility: sasflag never caused automatic equilibrium resplining, so qhigh probably shouldn't either. qlow, on the other hand, should force equilibrium resplining since it is a new function and someone using it would expect it to have an effect on the calculation

StuartBenjamin committed:
Adding updates from adding_MRE_terms to this branch

StuartBenjamin committed:
removing an old comment that seems a little confusing (to me at least)

StuartBenjamin committed:
Bugfix: sasflag was cutting an extra surface off after resplining due to a tiny discrepancy in qmax and the sasflag cutoff. Now after resplining, sasflag is set to false to avoid this bug.

StuartBenjamin commented:
Pause on this - I found a rare case where qhigh overrode sasflag in RDCON but sasflag overrode qhigh in STRIDE (will get to the bottom of it)

logan-nc commented:
Ok. This looks good to me otherwise. 

One iffy thing is that you are introducing unique truncation behavior to RDCON with the reform based on qlow. DCON and STRIDE to not automatically do this as far as I know. The motivation for the reform was that bad GSE in the edge could sometimes result in dirtier spline ringing further in, so reforming with the lower psihigh resulted in cleaner equilibrium descriptions. There was no such issue with the equilibrium in the core. Ideally, we'd keep all 3 codes as consistent as possible.

Skimming the repo, I don't see anywhere where the `sas_flag` impacted the galerkin gridding and so I would guess the qlow wouldn't either.... but I could easily be wrong on that. I guess reforming is always safest - should we do it in the others? Now that I look at it, I am scared the others could have a bug related to the `sing` type having more `msing` than it should if `qlow` cuts out a surface. It looks like `sing_find` doesn't know about `qlow` and I could imagine a bug where the integrator starting above the first `ising` could through things out of sync. I don't **think** there is a bug, but the only way to check is to reform and compare the results. 

*Request*: As part of this PR, can you run RDCON, STRIDE, and DCON with a qlow that moves psilow up past at least one singular surface and check if reforming the spline with lower bounds changes the result? If yes for any, then lets force reform for all (yay for consistency). If no for all, then we don't need to force it here.

StuartBenjamin committed:
merging new develop vaccum updates
Merge remote-tracking branch 'origin/develop' into standardising-truncation-of-rdcon-and-stride

StuartBenjamin committed:
making sing_find consistent with stride; exit when search exceeds qlim

StuartBenjamin committed:
making free.f vacuum call consistent with stride; rdcon now uses qlim, not qa when defining boundary for free_write_msc in  free_get_wvac

StuartBenjamin committed:
forcing a resplining of the equilibrium based on truncation logic (reasoning: rdcon's gal.f uses psihigh, psilow and the sq spline to define the domain of the galerkin elements... if we don't respline, redefining psihighm psilow & sq, then the truncation logic won't be used in the delta prime calculation)

StuartBenjamin committed:
only call sing_lim once

StuartBenjamin commented:
Hi Nik, so STRIDE automatically re-splines every time - there is an explanation of this straight from inside stride.f - 
`    ! Unlike DCON, we force a resplining.                                       `
`    ! The equil_out_qfind propogates to sing_find and then to the               `
`    ! parallelized intervals in a complicated web. Thus, it is not              `
`    ! sufficient to simply set the axisPsi to "start" the ODE somewhere new.    `
After more digging I found the rdcon deltaprime calculation in gal.f sets its domain based on psihigh, psilow and the sq spline (not qlim etc), so resplining should be done every time in rdcon as well. resplining isn't necessary in dcon, as you know, because the integration bounds are set by qlim etc.

Regarding qlow in stride, it basically re-sets psilow and then after resplining sing_find works fine. I just copied this across to rdcon.

I've ran two tests comparing the delta primes of rdcon and stride with this latest version, changing qlow to either include or exclude a rational surface, and qhigh to cycle through a couple rational surfaces. Here are printouts of the results:

<img width="1060" height="178" alt="Screenshot 2025-09-15 at 3 52 43 PM" src="https://github.com/user-attachments/assets/4c06a4be-9820-4cb3-960d-a181bb62adb7" />
<img width="741" height="128" alt="Screenshot 2025-09-15 at 3 54 39 PM" src="https://github.com/user-attachments/assets/cc10176c-ffd2-4598-8c13-5ea8d9800eb8" />

logan-nc reviewed (approved):
Ok. Based on the discussions recorded in this PR, I approve this top level logic and consistency with STRIDE. 

An open issue is inconsistency between DCON and STRIDE/RDCON... but that can be left for another PR for now.

I'm out on vacation until 9/29... @StuartBenjamin please wait for review approval from @matt-pharr and merge once you have it. TY.

matt-pharr committed:
RDCON - MINOR - Removed redundant spline eval, updated other example input files with new truncation variables

matt-pharr reviewed (commented):
Looks good! Looked through all the lines. I have not done any test runs yet, however. I also updated the other input files with the new variables.

StuartBenjamin commented:
Daniel gave the go ahead, and matt ran the default test cases (no seg fault), so I'm merging
